### PR TITLE
fix(kafka sink): retry messages that result in kafka policy violations

### PR DESCRIPTION
**Pull request by PriceHiller**
## Summary

Copied from the body of the commit:

> Problem: Some messages were getting dropped by Vector due to Kafka throwing `PolicyViolation` errors. These should be retried as a policy can be as simple as a more aggressive rate limit.
>
> ---
>
> Solution: Retry any messages that had the `RDKafkaErrorCode::PolicyViolation` error.
>
> ---
>
> Note: A dynamic back off may be better, as there may be a rate limit out there that still needs more than 100ms to back off on requests.
>
> ---
>
> See the original issue at https://github.com/vectordotdev/vector/issues/22026
>
> Closes https://github.com/vectordotdev/vector/issues/22026

Ensures Kafka retries messages that error out with `RDKafkaErrorCode::PolicyViolation`.

## Change Type

- [x] Bug fix
- [ ] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?

- [ ] Yes
- [x] No

## How did you test this PR?

- `cargo test --no-default-features --features "sinks-console,sinks-kafka,kafka-integration-tests" -- --skip 'sinks::kafka::tests::integration_test' 'sinks::kafka'`
- `make test-integration-kafka`
  - **NOTE**: I am facing a few errors when running the integration tests both with and without this patch. Are they currently broken by any chance?
    Has to do with `sources::kafka::integration_test::drains_acknowledgements_during_rebalance_default_assignments`

## Does this PR include user facing changes?

- [ ] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [x] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist

- [x] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- [x] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
      run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

- Closes: https://github.com/vectordotdev/vector/issues/22026

### CODE REVIEW

**Comment by PriceHiller** [comment]
Realized I did _not_ need to include that changelog entry for this PR.

**Review by jszwedko — COMMENTED** [review]
Actually could we add a changelog entry here describing the fix for users? See https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md

**Comment by PriceHiller** [comment]
> Actually could we add a changelog entry here describing the fix for users? See https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md

Reincluded the changelog entry.

File: changelog.d/22026-retry-kafka-policy-violations.fix.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Retry Kafka messages that error with `RDKafkaErrorCode::PolicyViolation` so messages are not lost.
```
Comment (jszwedko) [inline changelog.d/22026-retry-kafka-policy-violations.fix.md:1]:
```suggestion
Retry Kafka messages that error with a policy violation so messages are not lost.
```

**Review by jszwedko — APPROVED** [review]
Thanks @PriceHiller !

**Pull request merged**